### PR TITLE
Validate unparsed attributes and subentities in launch_xml and launch_yaml

### PR DESCRIPTION
**Pull request by ivanpauno**
Fixes https://github.com/ros2/launch/issues/338.

Adds three methods to `launch.frontend.Entity`:
- `assert_no_children(self)`
- `assert_subentity_types(self, types: Iterable[str])`
- `assert_attribute_names(self, names: Iterable[str])`

It might make the `parse` methods a bit more verbose, but it's not hard to use.
I have modified two parsing functions to use this, and it has already caught a preexisting error!

There might be some other ways to achieve the same, so I would like to get feedback about the approach before applying this more broadly.

### CODE REVIEW

**Review by jacobperron — COMMENTED** [review]
I think it would be nicer to have children and attributes explicitly declared versus calling separate assert functions. For example, by extending the `expose_action` decorator:

```python
@expose_action(
    'executable',
    allowed_attributes=(
        'cmd',
        'cwd',
        'name',
        'launch-prefix',
        'output',
        'respawn',
        'respawn_delay',
        'shell',
        'additional_env',
        'if',
        'unless'
    ),  
    allowed_children=('env', ))
class ExecuteProcess(Action):
    ...
```
If `allowed_*` is empty (or `None`), then implicitly none are allowed. This way we'll expose bugs in user launch scripts right away; the downside being that we'll break a lot of existing user code, but maybe we could tick-tock the feature, emitting a warning if they haven't explictly declared attributes/children they are accessing.

It might also be nice to have a flag to allow **any** attributes or children. I'm not sure what that would look like.

I don't know the details of the frontend implementation, so I'm not sure how feasible what I proposed is.

**Comment by ivanpauno** [comment]
> I think it would be nicer to have children and attributes explicitly declared versus calling separate assert functions. For example, by extending the expose_action decorator:

I like that one.
I will still need the `assert` functions in `launch.frontend.Entity`, but those can be called directly from the description in the decorator.

I will still need to distinguish between `allow_subentities` (e.g. env in `ExecuteProcess`), and allow children (e.g. any action in `GroupAction`).

File: launch/launch/actions/execute_process.py [hunk 1]
```diff
@@ -324,6 +324,12 @@ def parse(
         :param: ignore A list of arguments that should be ignored while parsing.
             Intended for code reuse in derived classes (e.g.: launch_ros.actions.Node).
         """
+        if ignore is None:
+            entity.assert_attribute_names((
+                'cmd', 'cwd', 'name', 'launch-prefix', 'output', 'respawn', 'respawn_delay', 'shell',
+                'additional_env', 'if', 'unless'
```
Comment (hidmic) [inline launch/launch/actions/execute_process.py:330]: @ivanpauno hmm, do we have to explicitly list base classes' attributes? What's going to happen with the assertions performed by another class that derives from this one?

File: launch_xml/launch_xml/entity.py [hunk 2]
```diff
@@ -55,6 +56,28 @@ def children(self) -> List['Entity']:
...
+                f'but the following were found: {tuple(item.tag for item in self.__xml_element)}'
+            )
+
+    def assert_subentity_types(self, types: Iterable[str]):
+        for item in self.__xml_element:
+            if item.tag not in types:
+                raise ValueError(
+                    f'Found subentity of type `{item.tag}` in a `{self.__xml_element.tag}`, '
+                    f'which expects only the following subentities: {types}'
+                )
```
Comment (hidmic) [inline launch_xml/launch_xml/entity.py:72]: @ivanpauno hmm, do we need this distinction between children and subentities? IIRC all of them are (sub)entities, whether we require them to be of an specific type or not.

**Review by hidmic — COMMENTED** [review]
First of all, to put things in context, I think that, to solve #338, we have two problems at hand: (1.) to enforce an schema as we parse, and (2.) to correlate any errors with the given input. To complain about a bad tag _somewhere_ in, say, a three levels deep launch file hierarchy is of little use.

Now, this patch addresses the first problem. I agree with @jacobperron's suggestion, but based on my observations, @ivanpauno I (personally) think it's time to go back to that old premise about frontends being standardized 1-to-1 mappings to their Python implementation. 

I'd suggest we use introspection from within our decorators to deduce what attributes and children should be based on `__init__` arguments (along with their annotated types), for the decorated class and its base classes, recursively. As bonus, we get to enforce that 1-to-1 mapping by the same price.

File: launch_yaml/test/launch_yaml/test_executable.py [hunk 3]
```diff
@@ -32,7 +32,7 @@ def test_executable():
                 name: my_ls
                 shell: true
                 output: log
-                launch_prefix: $(env LAUNCH_PREFIX)
+                'launch-prefix': $(env LAUNCH_PREFIX '')
```
[the remark this answers is not in the document]
Comment (ivanpauno) [reply launch_yaml/test/launch_yaml/test_executable.py:35]: The correct name is `launch-prefix` (not `launch_prefix`), and after adding the assert functions this test wasn't passing anymore.

File: launch/launch/actions/execute_process.py (hunk 1 above)
Comment (ivanpauno) [reply launch/launch/actions/execute_process.py:330]: That's why I used the trick `if ignore is None:`.
Not a beauty, but I don't know how to do better.

File: launch_xml/launch_xml/entity.py (hunk 2 above)
Comment (ivanpauno) [reply launch_xml/launch_xml/entity.py:72]: They aren't exactly the same in launch yaml AFAIR:

```yaml
launch:
  group:
    - executable:
      env:
        - ...
    - executable: ...
```

**Comment by ivanpauno** [comment]
> I (personally) think it's time to go back to that old premise about frontends being standardized 1-to-1 mappings to their Python implementation.
>
> I'd suggest we use the full power of Python's introspection machinery from within our decorators to deduce which attributes and children are valid from __init__ arguments (along with their annotated types) for the decorated class and its base classes, recursively. As a bonus, we get to enforce that 1-to-1 mapping.

I don't disagree, but we aren't mapping them 1-to-1 right now, and start doing that would be a breaking change.
In the case of a class with base classes the mechanism isn't obvious (e.g. node accepts different arguments that executable, but some are directly forwarded to the base class).
In that last case, the decorator will still need some user input (a solution to this is to never use `kwargs`, and always forward arguments explicitly in constructors).

We could still add a more intelligent mechanism that deduce the attribute names from the constructor arguments in the future, I think this improvement wouldn't block that (specially if following @jacobperron suggestion of providing the valid attributes and sub-entities in the decorator).

**Comment by ivanpauno** [comment]
It's also not obvious how the decorator would guess what's an attribute (name, executable, plugin, etc), what are children subentities (nested actions in group_action) and what are complex attributes (env, param, etc). Maybe there could be some heuristic based on type annotations, but I'm not sure if that will always work,

**Comment by hidmic** [comment]
> I don't disagree, but we aren't mapping them 1-to-1 right now, and start doing that would be a breaking change.

Hmm, aren't we? IIRC that's what we strived for. I'd suggest we don't delay in correcting the deviation. User provided aliases can aid deprecation cycles.

> We could still add a more intelligent mechanism that deduce the attribute names from the constructor arguments in the future, I think this improvement wouldn't block that

I agree schema deduction is not strictly necessary. To be clear, I meant it as a proposal to solve the problem. I'm sure there are other solutions, perhaps better ones. We can discuss them. But that is not to say I agree with this patch being an improvement as long as it requires replicating information across inheritance chains like this.

> In the case of a class with base classes the mechanism isn't obvious (e.g. node accepts different arguments that executable, but some are directly forwarded to the base class).

One solution to that problem (and perhaps an alternative, explicit, solution to the problem w/o using introspection) is to progressively tag input (at least for validation purposes) as it is accessed. If by the time you've processed an entire subtree, a part of it remains unacknowledged, then you know the system didn't recognize it.

> the decorator will still need some user input

This is correct. The canonical example being arbitrary children mapped to an specific argument. I'm not against user input, but against introducing unnecessary coupling in the system.

> It's also not obvious how the decorator would guess what's an attribute (name, executable, plugin, etc), what are children subentities (nested actions in group_action) and what are complex attributes (env, param, etc). Maybe there could be some heuristic based on type annotations, but I'm not sure if that will always work,

Once it is clear which (and if an) argument maps to all nested entities (which IIRC cannot be coexist with complex attributes), you can deduce the all attributes at first depth from type annotations. Of course, it cannot deduce the structure of complex attributes -- whose shape may be the result of a parsing function's implementation e.g. nested dictionaries.

**Comment by ivanpauno** [comment]
> One solution to that problem (and perhaps an alternative, explicit, solution to the problem w/o using introspection) is to progressively tag input (at least for validation purposes) as it is accessed. If by the time you've processed an entire subtree, a part of it remains unacknowledged, then you know the system didn't recognize it.

Yeah, I thought about that one.
I will try to implement something like that and see how it works.

**Comment by ivanpauno** [comment]
I have updated the PR with the approach suggested above, good news is that we don't need to update any parsing function.

File: launch/launch/frontend/expose.py [hunk 4]
```diff
@@ -98,7 +98,15 @@ def expose_impl_decorator(exposed):
                     name
                 )
             )
-        parse_methods_map[name] = found_parse_method
+        if exposed_type == 'action':
+            # For actions, validate that the user didn't provide unkown attributes or children
```
Comment (jacobperron) [inline launch/launch/frontend/expose.py:102]:
```suggestion
            # For actions, validate that the user didn't provide unknown attributes or children
```

File: launch_xml/launch_xml/entity.py [hunk 5]
```diff
@@ -53,8 +55,23 @@ def parent(self) -> Optional['Entity']:
...
         """Get the Entity's children."""
+        self.__read_children = {item.tag for item in self.__xml_element}
         return [Entity(item) for item in self.__xml_element]
 
+    def assert_entity_complely_parsed(self):
+        unparsed_nested_tags = {item.tag for item in self.__xml_element} - self.__read_children
+        if unparsed_nested_tags:
+            raise ValueError(
+                f'Found the following nested tags of `{self.__xml_element.tag}` that were not '
+                f'parsed: {unparsed_nested_tags}'
```
Comment (jacobperron) [inline launch_xml/launch_xml/entity.py:66]: Tiny nitpick: from a user perspective, I don't know if "parsed" is the right word. I suggest rewording to something like,  `"Unexpected nested tag(s) found in '{self.__xml_element.tag}': {unparsed_nested_tags}"`

File: launch_xml/launch_xml/entity.py [hunk 6]
```diff
@@ -53,8 +55,23 @@ def parent(self) -> Optional['Entity']:
...
+            )
+        unparsed_attributes = set(self.__xml_element.attrib.keys()) - self.__read_attributes
+        if unparsed_attributes:
+            raise ValueError(
+                f'Found the following attributes of `{self.__xml_element.tag}` that were not '
+                f'parsed: {unparsed_attributes}'
```
Comment (jacobperron) [inline launch_xml/launch_xml/entity.py:72]: Same nitpick here.

File: launch_yaml/launch_yaml/entity.py [hunk 7]
```diff
@@ -54,24 +56,46 @@ def parent(self) -> Optional['Entity']:
...
+        if not isinstance(self.__element, (dict, list)):
+            raise TypeError(
+                f'Expected a dict or list, got {type(self.element)}:'
+                f'\n---\n{self.__element}\n---'
+            )
         if isinstance(self.__element, dict):
             if 'children' not in self.__element:
-                raise KeyError('Missing `children` key in Entity {}'.format(self.__type_name))
+                raise ValueError(
+                    f'Expected entity {self.__type_name} to have children entities.'
```
Comment (jacobperron) [inline launch_yaml/launch_yaml/entity.py:68]:
```suggestion
                    f"Expected entity '{self.__type_name}' to have children entities."
```

File: launch/launch/frontend/expose.py [hunk 8]
```diff
@@ -98,7 +98,15 @@ def expose_impl_decorator(exposed):
                     name
                 )
             )
-        parse_methods_map[name] = found_parse_method
+        if exposed_type == 'action':
```
Comment (jacobperron) [inline launch/launch/frontend/expose.py:101]: Do we need to have this condition on the `exposed_type`? According the docstring, `exposed_type` is only used for clearer error logging.

**Review by jacobperron — COMMENTED** [review]
Generally looks good. I've left some minor comments

File: launch/launch/frontend/expose.py (hunk 4 above)
Comment (ivanpauno) [reply launch/launch/frontend/expose.py]: Fixed in https://github.com/ros2/launch/pull/468/commits/09871dd4925a29a27ac8f6b67a7add84f120d7c8.

File: launch_xml/launch_xml/entity.py (hunk 5 above)
Comment (ivanpauno) [reply launch_xml/launch_xml/entity.py]: Improved this one and similar error messages in https://github.com/ros2/launch/pull/468/commits/6ad4a3a47e0b50ed9d7ea581a159ccbffec821df.

File: launch/launch/frontend/expose.py (hunk 8 above)
Comment (ivanpauno) [reply launch/launch/frontend/expose.py:101]: There are two kinds of parsing functions, one for substitutions and one for actions.
The wrapper only apply to actions (because substitutions parsing functions are different).

Comment (ivanpauno) [reply launch/launch/frontend/expose.py:100]: Removed the old comment in https://github.com/ros2/launch/pull/468/commits/716e15d729495d6b5e3211cf585600eebe8cdc28.

**Review by jacobperron — APPROVED** [review]
LGTM. It'd be good to have @hidmic take a look too.

**Comment by ivanpauno** [comment]
CI, testing all downstream packages of `launch`:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12945/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7894/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10662/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12894/)

File: launch/launch/frontend/entity.py [hunk 9]
```diff
@@ -82,3 +82,12 @@ def get_attr(
             Only happens in frontend implementations that do type coercion.
         """
         raise NotImplementedError()
+
+    def assert_entity_complely_parsed(self):
```
Comment (hidmic) [inline launch/launch/frontend/entity.py:86]: @ivanpauno did you mean:
```suggestion
    def assert_entity_completely_parsed(self):
```
?

File: launch/launch/frontend/expose.py [hunk 10]
```diff
@@ -98,7 +97,15 @@ def expose_impl_decorator(exposed):
...
+            # For actions, validate that the user didn't provide unknown attributes or children
+            def wrapper(entity, parser):
```
Comment (hidmic) [inline launch/launch/frontend/expose.py:102]: @ivanpauno consider:
```suggestion
            @functools.wraps(found_parse_method)
            def wrapper(entity, parser):
```

File: launch_xml/launch_xml/entity.py [hunk 11]
```diff
@@ -53,8 +55,23 @@ def parent(self) -> Optional['Entity']:
     @property
     def children(self) -> List['Entity']:
         """Get the Entity's children."""
+        self.__read_children = {item.tag for item in self.__xml_element}
         return [Entity(item) for item in self.__xml_element]
 
+    def assert_entity_complely_parsed(self):
+        unparsed_nested_tags = {item.tag for item in self.__xml_element} - self.__read_children
```
Comment (hidmic) [inline launch_xml/launch_xml/entity.py:62]: @ivanpauno shouldn't we assert recursively? I'm thinking of nested structures that are parsed by the same parsing function e.g. nested `launch_ros.actions.Node` parameters. Thinking out loud, perhaps we can keep some form of context in the parser to avoid full traversals at each node of the tree, but that's arguably early optimization.

Comment (ivanpauno) [reply launch_xml/launch_xml/entity.py:62]: > shouldn't we assert recursively?

That doesn't sound like a good idea to me. I mean for things like `group` or `timer` it shouldn't be recursive, but for `executable` or `node` that could work.

>  I'm thinking of nested structures that are parsed by the same parsing function e.g. nested launch_ros.actions.Node parameters

Yeah that's true, maybe in those places we need to call `assert_entity_completely_parsed` manually (as `param`, `env`, etc don't have an explicit parsing function that can be wrapped).

File: launch/launch/frontend/entity.py (hunk 9 above)
Comment (ivanpauno) [reply launch/launch/frontend/entity.py]: ups, fixed in https://github.com/ros2/launch/pull/468/commits/dc8fc265f5ef4c7be2645210621d969d2dd0d7f8

File: launch/launch/frontend/expose.py (hunk 10 above)
Comment (ivanpauno) [reply launch/launch/frontend/expose.py:104]: Done in https://github.com/ros2/launch/pull/468/commits/7367cf9119dc6aa6f6edf1f556a8e1ad2d44233f

File: launch_xml/launch_xml/entity.py (hunk 11 above)
Comment (ivanpauno) [reply launch_xml/launch_xml/entity.py:62]: See https://github.com/ros2/launch/pull/468/commits/caf4b0cf338ee55218cc3c737d17c58d861f17af

Comment (ivanpauno) [reply launch_xml/launch_xml/entity.py:62]: and https://github.com/ros2/launch_ros/pull/198

File: launch_xml/test/launch_xml/test_arg.py [hunk 12]
```diff
@@ -34,3 +36,18 @@ def test_arg():
...
+        <launch>
+            <arg name="my_arg" whats_this="hello" default="asd" description="something"/>
+        </launch>
+        """
+    xml_file = textwrap.dedent(xml_file)
+    root_entity, parser = Parser.load(io.StringIO(xml_file))
+    with pytest.raises(ValueError) as excinfo:
+        parser.parse_description(root_entity)
+    assert '`arg`' in str(excinfo.value)
+    assert 'whats_this' in str(excinfo.value)
```
Comment (hidmic) [inline launch_xml/test/launch_xml/test_arg.py:53]: @ivanpauno can we add a test for a more complex (i.e. nested) description?

**Review by hidmic — APPROVED** [review]
LGTM pending green CI!

File: launch_xml/test/launch_xml/test_arg.py (hunk 12 above)
Comment (ivanpauno) [reply launch_xml/test/launch_xml/test_arg.py:53]: See https://github.com/ros2/launch/pull/468/commits/9dfb1a7a7cf575b4193c32b9753fbc958cdaddb9.

**Comment by ivanpauno** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12960/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7909/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10678/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12909/)

**Review by jacobperron — APPROVED** [review]

**Comment by adityapande-1995** [comment]
@Mergifyio backport foxy
